File: algo_for_cpp/118.pascals-triangle.py
#
# @lc app=leetcode id=118 lang=python3
#
# [118] Pascal's Triangle
#
from typing import List
# @lc code=start
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        if numRows == 0:
            return []
        if numRows == 1:
            return [[1]]
        if numRows == 2:
            return [[1], [1, 1]]
        if numRows >=3:
            res= [[1], [1,1]]
            row_last = [1,1]
            for i in range(3,numRows+1):
                row_now = [0]*i
                for j in range(i):
                    if j==0 or j== i-1:
                        row_now[j] = 1
                    else:
                        row_now[j] = row_last[j-1]+row_last[j]
                    
                row_last = row_now
                res.append(row_last)
            return res
if __name__ == "__main__":
    s = Solution()
    print(s.generate(5))


# @lc code=end

File: algo_for_cpp/118.pascals-triangle.cpp
/*
 * @lc app=leetcode id=118 lang=cpp
 *
 * [118] Pascal's Triangle
 */
#include <vector>
#include <iostream>
using namespace std;
// @lc code=start
class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        vector<vector<int>> res;
        vector<int> row_last={1,1};
        vector<int> row_now;
        if (numRows == 1){
            res.push_back({1});
            return res;
        }
        if (numRows == 2){
            res.push_back({1});
            res.push_back({1,1});
            return res;
        }
        if (numRows >=3){
            for (int i=0;i<numRows;i++){
                row_now.clear();
                for (int j=0;j<=i;j++){
                    if (j==0 || j==i){
                        row_now.push_back(1);
                                    }
                    else{
                        row_now.push_back(row_last[j]+row_last[j-1]);
                    }
                
                    }
                row_last = row_now;
                res.push_back(row_now);
                }
            return res;

            }
        }
    
};
// @lc code=end

